localstack/services/awslambda/invocation/runtime_executor_plugin.py

```diff
@@ -0,0 +1,111 @@
+
+LOG = logging.getLogger(__name__)
+
+
+class RuntimeExecutorPlugin(Plugin):
+    namespace = "localstack.lambda.runtime_executor"
+
+
+class RuntimeExecutor(ABC):
+    id: str
+    function_version: FunctionVersion
+
+    def __init__(
+        self, id: str, function_version: FunctionVersion, service_endpoint: ServiceEndpoint
+    ) -> None:
+        """
+        Runtime executor class responsible for executing a runtime in specific environment
+
+        :param id: ID string of the runtime executor
+        :param function_version: Function version to be executed
+        :param service_endpoint: Service endpoint for execution related callbacks
+        """
+        self.id = id
+        self.function_version = function_version
+
+    @abstractmethod
+    def start(self, env_vars: dict[str, str]) -> None:
+        """
+        Start the runtime executor with the given environment variables
+
+        :param env_vars:
+        :return:
+        """
+        pass
+
+    @abstractmethod
+    def stop(self) -> None:
+        """
+        Stop the runtime executor
+        """
+        pass
+
+    @abstractmethod
+    def get_address(self) -> str:
+        """
+        Get the address the runtime executor is available at for the LocalStack container.
+
+        :return: IP address or hostname of the execution environment
+        """
+        pass
+
+    @abstractmethod
+    def get_endpoint_from_executor(self) -> str:
+        """
+        Get the address of LocalStack the runtime execution environment can communicate with LocalStack
+
+        :return: IP address or hostname of LocalStack (from the view of the execution environment)
+        """
+        pass
+
+    @abstractmethod
+    def invoke(self, payload: dict[str, str]) -> None:
+        """
+        Send an invocation to the execution environment
+
+        :param payload: Invocation payload
+        """
+        pass
+
+    @staticmethod
+    @abstractmethod
+    def prepare_version(function_version) -> None:
+        """
+        Prepare a given function version to be executed.
+        Includes all the preparation work necessary for execution, short of starting anything
+
+        :param function_version: Function version to prepare
+        """
+        pass
+
+    @staticmethod
+    @abstractmethod
+    def cleanup_version(function_version):
+        """
+        Cleanup the version preparation for the given version.
+        Should cleanup preparation steps taken by prepare_version
+        :param function_version:
+        """
+        pass
+
+
+EXECUTOR_PLUGIN_MANAGER: PluginManager[Type[RuntimeExecutor]] = PluginManager(
+    RuntimeExecutorPlugin.namespace
+)
+
+
+def get_runtime_executor() -> Type[RuntimeExecutor]:
+    plugin_name = config.LAMBDA_RUNTIME_EXECUTOR or "docker"
+    if not EXECUTOR_PLUGIN_MANAGER.exists(plugin_name):
+        LOG.warning('Invalid specified plugin name %s. Falling back to "docker" runtime executor')
+        plugin_name = "docker"
```

Comment:
This seems to conflict with each other a bit since Line 107 already sets a default value for `plugin_name` so the `if` branch will never be entered here

Reply:
I think the `or "docker"` is more unnecessary here, because someone could specify and executor he either has no access to, or is just plain wrong, in this case the if would be called. The `or "docker"` is just here so no warning is printed if the option is just left out.

Reply:
Makes sense. In that case the format string seems to be missing its replacement value for `%s` though. We can leave the first `docker` default since as you've said it should at least avoid some of the warnings :man_shrugging:

Reply:
Ah, overlooked the replacement value. Fixed!